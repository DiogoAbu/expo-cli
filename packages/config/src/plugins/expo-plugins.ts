import { ConfigPlugin } from '../Plugin.types';
import * as IOSConfig from '../ios';
import { withPlugins } from './core-plugins';

/**
 * Config plugin to apply all of the custom Expo iOS config plugins we support by default.
 * TODO: In the future most of this should go into versioned packages like expo-facebook, expo-updates, etc...
 */
export const withExpoIOSPlugins: ConfigPlugin<{
  bundleIdentifier: string;
  expoUsername: string | null;
}> = (config, { bundleIdentifier, expoUsername }) => {
  // Set the bundle ID ahead of time.
  if (!config.expo.ios) config.expo.ios = {};
  config.expo.ios.bundleIdentifier = bundleIdentifier;

  return withPlugins(config, [
    [IOSConfig.BundleIdenitifer.withBundleIdentifier, { bundleIdentifier }],
    IOSConfig.Branch.withBranch,
    IOSConfig.Facebook.withFacebook,
    IOSConfig.Google.withGoogle,
    IOSConfig.Name.withDisplayName,
    // IOSConfig.Name.withName,
    IOSConfig.Orientation.withOrientation,
    IOSConfig.RequiresFullScreen.withRequiresFullScreen,
    IOSConfig.Scheme.withScheme,
    IOSConfig.UserInterfaceStyle.withUserInterfaceStyle,
    IOSConfig.UsesNonExemptEncryption.withUsesNonExemptEncryption,
    IOSConfig.Version.withBuildNumber,
    IOSConfig.Version.withVersion,
    IOSConfig.Google.withGoogleServicesFile,
    [IOSConfig.Updates.withUpdates, { expoUsername }],
    // Entitlements
    IOSConfig.Entitlements.withAppleSignInEntitlement,
    IOSConfig.Entitlements.withAccessesContactNotes,
    // TODO: We don't have a mechanism for getting the apple team id here yet
    [IOSConfig.Entitlements.withICloudEntitlement, { appleTeamId: 'TODO-GET-APPLE-TEAM-ID' }],
    IOSConfig.Entitlements.withAssociatedDomains,
    // XcodeProject
    IOSConfig.DeviceFamily.withDeviceFamily,
    IOSConfig.Locales.withLocales,
    // Dangerous
    IOSConfig.Icons.withIcons,
    IOSConfig.SplashScreen.withSplashScreen,
  ]);
};

/**
 * Config plugin to apply all of the custom Expo Android config plugins we support by default.
 * TODO: In the future most of this should go into versioned packages like expo-facebook, expo-updates, etc...
 */
export const withExpoAndroidPlugins: ConfigPlugin<unknown> = config => {
  return withPlugins(config, [
    // TODO: Support Android
  ]);
};
