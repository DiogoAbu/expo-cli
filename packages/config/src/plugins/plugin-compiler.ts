import { JSONObject } from '@expo/json-file';
import { IosPlist } from '@expo/xdl';
import { writeFile } from 'fs-extra';
import path from 'path';
import { XcodeProject } from 'xcode';

import {
  ConfigModifierPlugin,
  ExpoConfig,
  ExportedConfig,
  IOSPluginModifierProps,
  PluginConfig,
  PluginPlatform,
} from '../Config.types';
import * as WarningAggregator from '../WarningAggregator';
import { InfoPlist } from '../ios';
import { getEntitlementsPath } from '../ios/Entitlements';
import { getPbxproj, getProjectName } from '../ios/utils/Xcodeproj';
import { ensureArray, withInterceptedModifier } from './core-plugins';

export async function compilePluginsAsync(projectRoot: string, config: ExportedConfig) {
  config = applyIOSCoreModifiers(projectRoot, config);
  config = applyAndroidCoreModifiers(projectRoot, config);

  return await evalPluginsAsync(config, { projectRoot });
}

function resolveModifierResults(results: any, platformName: string, modifierName: string) {
  // If the results came from a modifier, they'd be in the form of [config, data].
  // Ensure the results are an array and omit the data since it should've been written by a data provider plugin.
  const ensuredResults = ensureArray(results)[0] as any;

  // Sanity check to help locate non compliant modifiers.
  if (!ensuredResults?.expo || !ensuredResults?.plugins) {
    throw new Error(
      `Modifier \`plugins.${platformName}.${modifierName}\` evaluated to an object that is not a valid project config. Instead got: ${JSON.stringify(
        ensuredResults
      )}`
    );
  }
  return ensuredResults;
}

function applyAndroidCoreModifiers(projectRoot: string, config: ExportedConfig): ExportedConfig {
  // TODO: Support android plugins
  return config;
}

function applyIOSCoreModifiers(projectRoot: string, config: ExportedConfig): ExportedConfig {
  const { iosProjectDirectory, entitlementsPath, supportingDirectory } = getIOSPaths(
    projectRoot,
    config.expo
  );

  // Append a rule to supply Info.plist data to plugins on `plugins.ios.infoPlist`
  config = withInterceptedModifier<IOSPluginModifierProps<InfoPlist>>(config, {
    platform: 'ios',
    modifier: 'infoPlist',
    async action(config, { nextModifier, ...props }) {
      let results: [ExportedConfig, IOSPluginModifierProps<JSONObject>] = [config, props];
      await IosPlist.modifyAsync(iosProjectDirectory, 'Info', async data => {
        // Apply all of the Info.plist values to the expo.ios.infoPlist object
        // TODO: Remove this in favor of just overwriting the Info.plist with the Expo object. This will enable people to actually remove values.
        if (!config.expo.ios) {
          config.expo.ios = {};
        }
        if (!config.expo.ios.infoPlist) {
          config.expo.ios.infoPlist = {};
        }

        config.expo.ios.infoPlist = {
          ...(data || {}),
          ...config.expo.ios.infoPlist,
        };

        results = await nextModifier(config, {
          ...props,
          data: config.expo.ios.infoPlist as InfoPlist,
        });
        resolveModifierResults(results, props.platform, props.modifier);
        return results[1].data;
      });
      await IosPlist.cleanBackupAsync(iosProjectDirectory, 'Info', false);
      return results;
    },
  });

  // Append a rule to supply Expo.plist data to plugins on `plugins.ios.expoPlist`
  config = withInterceptedModifier<IOSPluginModifierProps<JSONObject>>(config, {
    platform: 'ios',
    modifier: 'expoPlist',
    async action(config, { nextModifier, ...props }) {
      let results: [ExportedConfig, IOSPluginModifierProps<JSONObject>] = [config, props];
      try {
        await IosPlist.modifyAsync(supportingDirectory, 'Expo', async data => {
          results = await nextModifier(config, {
            ...props,
            data,
          });
          resolveModifierResults(results, props.platform, props.modifier);
          return results[1].data;
        });
      } catch (error) {
        WarningAggregator.addWarningIOS(
          'updates',
          'Expo.plist configuration could not be applied. You will need to create Expo.plist if it does not exist and add Updates configuration manually.',
          'https://docs.expo.io/bare/updating-your-app/#configuration-options'
        );
      } finally {
        await IosPlist.cleanBackupAsync(supportingDirectory, 'Expo', false);
      }
      return results;
    },
  });

  // Append a rule to supply .entitlements data to plugins on `plugins.ios.entitlements`
  config = withInterceptedModifier<IOSPluginModifierProps<JSONObject>>(config, {
    platform: 'ios',
    modifier: 'entitlements',
    async action(config, { nextModifier, ...props }) {
      let results: [ExportedConfig, IOSPluginModifierProps<JSONObject>] = [config, props];

      const directory = path.dirname(entitlementsPath);
      const filename = path.basename(entitlementsPath, 'plist');

      try {
        await IosPlist.modifyAsync(directory, filename, async data => {
          // Apply all of the .entitlements values to the expo.ios.entitlements object
          // TODO: Remove this in favor of just overwriting the .entitlements with the Expo object. This will enable people to actually remove values.
          if (!config.expo.ios) {
            config.expo.ios = {};
          }
          if (!config.expo.ios.entitlements) {
            config.expo.ios.entitlements = {};
          }

          config.expo.ios.entitlements = {
            ...(data || {}),
            ...config.expo.ios.entitlements,
          };

          results = await nextModifier(config, {
            ...props,
            data: config.expo.ios.entitlements as JSONObject,
          });
          resolveModifierResults(results, props.platform, props.modifier);
          return results[1].data;
        });
      } catch (error) {
        WarningAggregator.addWarningIOS(
          'entitlements',
          `${filename} configuration could not be applied.`
        );
      } finally {
        await IosPlist.cleanBackupAsync(directory, filename, false);
      }
      return results;
    },
  });

  // Append a rule to supply .xcodeproj data to plugins on `plugins.ios.xcodeproj`
  config = withInterceptedModifier<IOSPluginModifierProps<XcodeProject>>(config, {
    platform: 'ios',
    modifier: 'xcodeproj',
    async action(config, { nextModifier, ...props }) {
      const data = getPbxproj(projectRoot);
      const results = await nextModifier(config, {
        ...props,
        data,
      });
      resolveModifierResults(results, props.platform, props.modifier);
      const resultData = results[1].data;
      await writeFile(resultData.filepath, resultData.writeSync());
      return results;
    },
  });

  return config;
}

/**
 * A generic plugin compiler.
 *
 * @param config
 */
export async function evalPluginsAsync(
  config: ExportedConfig,
  props: { projectRoot: string }
): Promise<ExportedConfig> {
  for (const [platformName, platform] of Object.entries(config.plugins ?? ({} as PluginConfig))) {
    const platformProjectRoot = path.join(props.projectRoot, platformName);
    const projectName = platformName === 'ios' ? getProjectName(props.projectRoot) : undefined;

    for (const [modifier, plugin] of Object.entries(platform)) {
      const results = await (plugin as ConfigModifierPlugin<{
        projectRoot: string;
        platformProjectRoot: string;
        projectName?: string;
        modifier: string;
        platform: PluginPlatform;
      }>)(config, {
        ...props,
        projectName,
        platformProjectRoot,
        platform: platformName as PluginPlatform,
        modifier,
      });

      // Sanity check to help locate non compliant modifiers.
      config = resolveModifierResults(results, platformName, modifier);
    }
  }
  return config;
}

// TODO: come up with a better solution for using app.json expo.name in various places
function sanitizedName(name: string) {
  return name
    .replace(/[\W_]+/g, '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '');
}

// TODO: it's silly and kind of fragile that we look at app config to determine
// the ios project paths. Overall this function needs to be revamped, just a
// placeholder for now! Make this more robust when we support applying config
// at any time (currently it's only applied on eject).
function getIOSPaths(projectRoot: string, exp: ExpoConfig) {
  let projectName: string | null = null;

  // Attempt to get the current ios folder name (apply).
  try {
    projectName = getProjectName(projectRoot);
  } catch {
    // If no iOS project exists then create a new one (eject).
    projectName = exp.name;
    if (!projectName) {
      throw new Error('Your project needs a name in app.json/app.config.js.');
    }
    projectName = sanitizedName(projectName);
  }

  const iosProjectDirectory = path.join(projectRoot, 'ios', projectName);
  const iconPath = path.join(iosProjectDirectory, 'Assets.xcassets', 'AppIcon.appiconset');
  const supportingDirectory = path.join(iosProjectDirectory, 'Supporting');
  const entitlementsPath = getEntitlementsPath(projectRoot);
  return {
    projectName,
    supportingDirectory,
    iosProjectDirectory,
    entitlementsPath,
    iconPath,
  };
}
