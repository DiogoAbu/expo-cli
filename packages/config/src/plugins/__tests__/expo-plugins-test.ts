import fs from 'fs-extra';
import { vol } from 'memfs';

import { ExportedConfig } from '../../Config.types';
import { withExpoIOSPlugins } from '../expo-plugins';
import { compileModifiersAsync } from '../plugin-compiler';
import rnFixture from './fixtures/react-native-project';

jest.mock('fs');

describe(withExpoIOSPlugins, () => {
  const projectRoot = '/app';
  beforeEach(async () => {
    // Trick XDL Info.plist reading
    Object.defineProperty(process, 'platform', {
      value: 'not-darwin',
    });
    vol.fromJSON(rnFixture, projectRoot);
  });

  afterEach(() => {
    vol.reset();
  });

  it('prefers named keys over info.plist overrides', async () => {
    let config: ExportedConfig = {
      expo: {
        name: 'app',
        slug: '',
        ios: {
          config: {
            usesNonExemptEncryption: false,
          },
          infoPlist: {
            ITSAppUsesNonExemptEncryption: true,
          },
        },
      },
    };

    config = withExpoIOSPlugins(config, {
      bundleIdentifier: 'com.bacon.todo',
      expoUsername: 'bacon',
    });
    // Apply modifier
    config = await compileModifiersAsync('/app', config);
    // This should be false because ios.config.usesNonExemptEncryption is used in favor of ios.infoPlist.ITSAppUsesNonExemptEncryption
    expect(config.expo.ios?.infoPlist?.ITSAppUsesNonExemptEncryption).toBe(false);
  });

  it('compiles modifiers', async () => {
    let config: ExportedConfig = {
      expo: {
        name: 'app',
        slug: '',
        ios: {
          config: {
            usesNonExemptEncryption: true,
            branch: {
              apiKey: 'MY_BRANCH_KEY',
            },
            googleSignIn: {
              reservedClientId: 'GOOGLE_SIGN_IN_CLIENT_ID',
            },
          },
        },
      },
      modifiers: null,
    };

    config = withExpoIOSPlugins(config, {
      bundleIdentifier: 'com.bacon.todo',
      expoUsername: 'bacon',
    });

    // Apply modifier
    config = await compileModifiersAsync('/app', config);

    // App config should have been modified
    expect(config.expo.name).toBe('app');
    expect(config.expo.ios.infoPlist).toBeDefined();
    expect(config.expo.ios.entitlements).toBeDefined();
    // Google Sign In
    expect(
      config.expo.ios?.infoPlist?.CFBundleURLTypes?.find(({ CFBundleURLSchemes }) =>
        CFBundleURLSchemes.includes('GOOGLE_SIGN_IN_CLIENT_ID')
      )
    ).toBeDefined();
    // Branch
    expect(config.expo.ios?.infoPlist?.branch_key?.live).toBe('MY_BRANCH_KEY');

    // Shape
    expect(config.expo).toStrictEqual({});

    // Modifiers should all be functions
    expect(Object.values(config.modifiers.ios).every(value => typeof value === 'function')).toBe(
      true
    );

    // Test that the actual file was rewritten.
    const data = await fs.readFile('/app/ios/ReactNativeProject/Info.plist', 'utf8');
    expect(data).toMatch(/com.bacon.todo/);
  });
});
