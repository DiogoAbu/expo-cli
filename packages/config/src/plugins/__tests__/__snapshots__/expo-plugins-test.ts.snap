// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`withExpoIOSPlugins compiles modifiers 1`] = `
Object {
  "androidNavigationBar": Object {
    "backgroundColor": "#ff0000",
    "barStyle": "dark-content",
    "visible": "sticky-immersive",
  },
  "androidStatusBar": Object {
    "backgroundColor": "#000FFF",
    "barStyle": "light-content",
    "hidden": false,
    "translucent": true,
  },
  "appKey": "othermain",
  "backgroundColor": "orange",
  "description": "my app is great because it uses expo",
  "developmentClient": Object {
    "silentLaunch": true,
  },
  "facebookAdvertiserIDCollectionEnabled": true,
  "facebookAppId": "1234567890",
  "facebookAutoInitEnabled": true,
  "facebookAutoLogAppEventsEnabled": true,
  "facebookDisplayName": "my-fb-test-app",
  "facebookScheme": "fb1234567890",
  "githubUrl": "https://github.com/expo/expo",
  "ios": Object {
    "accessesContactNotes": true,
    "appStoreUrl": "https://itunes.apple.com/us/app/pillar-valley/id1336398804?ls=1&mt=8",
    "associatedDomains": Array [
      "applinks:https://pillarvalley.netlify.app",
    ],
    "backgroundColor": "#ff0000",
    "buildNumber": "6.5.0",
    "bundleIdentifier": "com.bacon.todo",
    "config": Object {
      "branch": Object {
        "apiKey": "MY_BRANCH_KEY",
      },
      "googleMapsApiKey": "TEST_googleMapsApiKey",
      "googleMobileAdsAppId": "TEST_googleMobileAdsAppId",
      "googleMobileAdsAutoInit": true,
      "googleSignIn": Object {
        "reservedClientId": "GOOGLE_SIGN_IN_CLIENT_ID",
      },
      "usesNonExemptEncryption": true,
    },
    "entitlements": Object {
      "com.apple.developer.applesignin": Array [
        "Default",
      ],
      "com.apple.developer.associated-domains": Array [
        "applinks:https://pillarvalley.netlify.app",
      ],
      "com.apple.developer.contacts.notes": true,
      "foo": "bar",
    },
    "googleServicesFile": "./config/GoogleService-Info.plist",
    "infoPlist": Object {
      "CFBundleDevelopmentRegion": "en",
      "CFBundleDisplayName": "my cool app",
      "CFBundleExecutable": "$(EXECUTABLE_NAME)",
      "CFBundleIdentifier": "$(PRODUCT_BUNDLE_IDENTIFIER)",
      "CFBundleInfoDictionaryVersion": "6.0",
      "CFBundleName": "$(PRODUCT_NAME)",
      "CFBundlePackageType": "APPL",
      "CFBundleShortVersionString": "1.0.0",
      "CFBundleSignature": "????",
      "CFBundleURLTypes": Array [
        Object {
          "CFBundleURLSchemes": Array [
            "my-app-redirect",
            "com.bacon.todo",
          ],
        },
        Object {
          "CFBundleURLSchemes": Array [
            "GOOGLE_SIGN_IN_CLIENT_ID",
          ],
        },
        Object {
          "CFBundleURLSchemes": Array [
            "fb1234567890",
          ],
        },
      ],
      "CFBundleVersion": "6.5.0",
      "FacebookAdvertiserIDCollectionEnabled": true,
      "FacebookAppID": "1234567890",
      "FacebookAutoInitEnabled": true,
      "FacebookAutoLogAppEventsEnabled": true,
      "FacebookDisplayName": "my-fb-test-app",
      "GADApplicationIdentifier": "TEST_googleMobileAdsAppId",
      "GMSApiKey": "TEST_googleMapsApiKey",
      "ITSAppUsesNonExemptEncryption": true,
      "LSApplicationQueriesSchemes": Array [
        "fbapi",
        "fb-messenger-api",
        "fbauth2",
        "fbshareextension",
      ],
      "LSRequiresIPhoneOS": true,
      "NSAppTransportSecurity": Object {
        "NSAllowsArbitraryLoads": true,
        "NSExceptionDomains": Object {
          "localhost": Object {
            "NSExceptionAllowsInsecureHTTPLoads": true,
          },
        },
      },
      "NSLocationWhenInUseUsageDescription": "",
      "UILaunchStoryboardName": "LaunchScreen",
      "UIRequiredDeviceCapabilities": Array [
        "armv7",
      ],
      "UIRequiresFullScreen": true,
      "UISupportedInterfaceOrientations": Array [
        "UIInterfaceOrientationPortrait",
        "UIInterfaceOrientationPortraitUpsideDown",
        "UIInterfaceOrientationLandscapeLeft",
        "UIInterfaceOrientationLandscapeRight",
      ],
      "UIUserInterfaceStyle": "Automatic",
      "UIViewControllerBasedStatusBarAppearance": false,
      "bar": Object {
        "val": Array [
          "foo",
        ],
      },
      "branch_key": Object {
        "live": "MY_BRANCH_KEY",
      },
    },
    "isTabletOnly": false,
    "merchantId": "TEST_MERCHANT_ID",
    "publishBundlePath": "./ios-dist",
    "requireFullScreen": true,
    "supportsTablet": true,
    "userInterfaceStyle": "automatic",
    "usesAppleSignIn": true,
    "usesIcloudStorage": true,
  },
  "locales": Object {
    "en": "./locales/en-US.json",
    "es": Object {
      "foo": "el bar",
    },
  },
  "name": "my cool app",
  "notification": Object {
    "androidCollapsedTitle": "#{unread_notifications} new interactions",
    "androidMode": "collapse",
    "color": "green",
    "icon": "./icons/notification-icon.png",
    "iosDisplayInForeground": true,
  },
  "orientation": "default",
  "packagerOpts": Object {
    "extraThing": true,
  },
  "platforms": Array [
    "android",
    "ios",
    "web",
  ],
  "primaryColor": "#fff000",
  "scheme": "my-app-redirect",
  "slug": "mycoolapp",
  "updates": Object {
    "checkAutomatically": "ON_ERROR_RECOVERY",
    "enabled": true,
    "fallbackToCacheTimeout": 650,
  },
  "userInterfaceStyle": "dark",
  "version": "1.0.0",
}
`;
